### PR TITLE
fix: don't revert gas refund logic when transaction reverted

### PR DESCRIPTION
**Pull request by yihuang**
## Description

Problem: When there are hooks and the transaction reverted, gas refund logic is reverted too.
Solution: Don't revert the gas refund logic

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: tests/solidity/suites/exception/contracts/TestRevert.sol [hunk 1]
```diff
@@ -32,6 +32,9 @@ contract TestRevert {
     function set(uint256 input) public {
         state.force_set(input);
     }
+    function direct_set(uint256 input) public {
```
Comment (yihuang) [inline tests/solidity/suites/exception/contracts/TestRevert.sol:35]:
```suggestion
    // will revert when imput>=10
    function direct_set(uint256 input) public {
```

File: x/evm/handler_test.go [hunk 2]
```diff
@@ -525,3 +525,79 @@ func (suite *EvmTestSuite) TestErrorWhenDeployContract() {
...
+	suite.Require().NoError(err)
+
+	res, err := k.EthereumTx(sdk.WrapSDKContext(suite.ctx), tx)
+	suite.Require().NoError(err)
+	suite.Require().True(res.Failed())
+
+	after := k.GetBalance(suite.from)
+
+	suite.Require().Equal(uint64(40745), res.GasUsed)
+	// suite.Require().Equal(big.NewInt(41000), new(big.Int).Sub(before, after))
```
Comment (yihuang) [inline x/evm/handler_test.go:591]:
```suggestion
	// check gas refund works
```

File: x/evm/types/compiled_contract.go [hunk 3]
```diff
@@ -68,6 +68,12 @@ func (s *CompiledContract) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// CompiledContract2 support json format of newer solc
+type CompiledContract2 struct {
```
Comment (fedekunze) [inline x/evm/types/compiled_contract.go:72]: rename this struct

File: x/evm/handler_test.go [hunk 4]
```diff
@@ -525,3 +525,79 @@ func (suite *EvmTestSuite) TestErrorWhenDeployContract() {
...
+		nil,
+		true,
+	)
+	rsp, err := k.ApplyMessage(msg, nil, true)
+	suite.Require().NoError(err)
+	suite.Require().False(rsp.Failed())
+	return crypto.CreateAddress(suite.from, nonce)
+}
+
+func (suite *EvmTestSuite) TestGasWhenReverted() {
```
Comment (fedekunze) [inline x/evm/handler_test.go:551]: can we refactor this test? it's not maintainable

**Review by fedekunze — CHANGES_REQUESTED** [review]

File: x/evm/handler_test.go (hunk 4 above)
Comment (yihuang) [reply x/evm/handler_test.go]: I realize we don't need a special contract to reproduce the contract revert, so I changed it to the existing erc20 contract, is it good now?

File: x/evm/types/compiled_contract.go (hunk 3 above)
Comment (yihuang) [reply x/evm/types/compiled_contract.go]: remoevd.

File: x/evm/keeper/state_transition.go [hunk 5]
```diff
@@ -241,6 +235,15 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 		}
 	}
 
+	// change to original context
+	k.WithContext(ctx)
+
+	// refund gas according to Ethereum gas accounting rules.
+	err = k.RefundGas(msg, msg.Gas()-res.GasUsed, cfg.Params.EvmDenom)
+	if err != nil {
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:243]:
```suggestion
	if err: = k.RefundGas(msg, msg.Gas()-res.GasUsed, cfg.Params.EvmDenom); err != nil {
```

**Review by fedekunze — APPROVED** [review]

**Pull request merged**